File: 5/sol.py
from collections import Counter

def parse_point(s):
    return tuple(int(c) for c in s.split(","))

def parse_line(line):
    return [parse_point(p) for p in line.split(" -> ")]

def delta(n1, n2):
    if (s := n2 - n1) > 0:
        return 1    
    elif s == 0:
        return 0    
    return -1

def span(p1, p2):
    sx, sy = p1 
    ex, ey = p2
    dx = delta(sx, ex)
    dy = delta(sy, ey)
    while p1 != p2:
        yield p1    
        p1 = (p1[0] + dx, p1[1] + dy)
    yield p1

def count_points(segments):
    c = Counter()
    for p1, p2 in segments:
        for p in span(p1, p2):
            c[p] += 1
    return c

with open("input.txt") as f:
    txt = f.read()
    segments = [parse_line(l) for l in txt.splitlines()]

def part1(segments):
    segments = [s for s in segments if s[0][0] == s[1][0] or s[0][1] == s[1][1]]
    c = count_points(segments)
    return sum(1 for v in c.values() if v > 1)
    
def part2(segments):
    c = count_points(segments)
    return sum(1 for v in c.values() if v > 1)

print(part1(segments))
print(part2(segments))

File: 5/sol.clj
(require '[clojure.string :as string])

(defn get-lines [file]
  (with-open [rdr (clojure.java.io/reader file)]
    (doall (line-seq rdr))))

(defn parse-point [s]
  (map #(Integer/parseInt %) (string/split s #",")))

(defn parse-segment [line]
  (flatten
    (map
      parse-point
      (string/split line #" -> "))))

(defn span [sx sy ex ey]
  (let [dx (delta sx ex)
        dy (delta sy ey)]
    (take-while #(not (= (list (+ ex dx) (+ ey dy)) %))
                (iterate
                  #(list (+ (nth % 0) dx) (+ (nth % 1) dy))
                  (list sx sy)))))

(defn into-counter [c point]
  (let [[sx sy ex ey] point]
    (reduce
      #(if
         (contains? %1 %2)
         (update %1 %2 inc)
         (assoc %1 %2 1))
      c
      (span sx sy ex ey))))

(defn points-counter [segments]
  (reduce
    #(into-counter %1 %2)
    {}
    segments))

(defn delta [n1 n2]
  (cond
    (pos? (- n2 n1)) 1 
    (zero? (- n2 n1)) 0
    :else -1))


(defn horizontal-and-vertical [segments]
  (filter
    #(let
       [[sx sy ex ey] %]
       (or (== sx ex) (== sy ey)))
    segments))


(defn more-than-one-time [counter]
  (filter #(let [[_, v] %]
             (>= v 2))
          counter))

(let [segments (map parse-segment (get-lines "input.txt"))]
  (do (println
        (count
          (more-than-one-time (points-counter (horizontal-and-vertical segments)))))
      (println
        (count
          (more-than-one-time (points-counter segments))))))
